feat: Add turing helm chart

krithika369 opened the pull request:
# Motivation
This PR is the first step to add the turing helm chart from the [caraml-dev/turing](https://github.com/caraml-dev/turing) repo ([ref](https://github.com/caraml-dev/turing/tree/main/infra/charts/turing)).

If MLP/Merlin are already installed, the Turing chart can be installed as follows:

```sh
helm install turing ./charts/turing \
  --set uiConfig.authConfig.oauthClientId=${OAUTH_CLIENT_ID} \
  -f charts/turing/values-global.yaml \
  -f charts/turing/ci/ci-values.yaml \
  --set config.MLPConfig.MLPEncryptionKey="secret-encryption" \
  --set turing-postgresql.persistence.enabled=false \
  --timeout=5m \
  --wait
```

The Turing chart can also be used to install Merlin and MLP  as dependencies (MLP can be a top-level dependency of Turing or nested dependency under Merlin).

# Modification
* `charts/turing/` - Add the Turing chart
* `charts/turing/tests/` - Add some unit tests for the deployment and secret specs
* Fix typo in the codebase from `secret-encyrption` to `secret-encryption`
* `scripts/generate-cluster-creds.sh` - Update to include Turing values files for writing the cluster credentials
* `charts/caraml/ci/ci-values.yaml`, `charts/merlin/ci/ci-values.yaml` - Removing the whitespaces that get auto-removed after the `scripts/generate-cluster-creds.sh` is run.

# Checklist
- [x] Chart version bumped
- [x] README.md updated

# Future Improvements
* Add `minio`, `mlflow` and `knative` as top-level dependencies on the `caraml` umbrella chart which get included if either Merlin or Turing are enabled. (Currently, Turing depends on Merlin but this dependency will be removed in the future.)
* Let Merlin / Turing directly depend on the Encryption secret created for MLP rather than create duplicate secrets.

krithika369 commented on charts/merlin/local_development.md:
```diff
@@ -45,7 +45,7 @@ This will generate a cluster-credentials.json file under `charts/merlin/files` d
 
 helm install merlin caraml/merlin \
-  --set merlin.oauthClientID=${OAUTH_CLIENT_ID} \
+  --set ui.oauthClientID=${OAUTH_CLIENT_ID} \
```
`oauthClientID` is under the `ui` section.

mbruner commented on charts/turing/templates/turing-service-account.yaml:
```diff
@@ -0,0 +1,8 @@
+{{- if .Values.clusterConfig.useInClusterConfig -}}
```
We should be able to:
- have a service account not only for in-cluster deployments
- set annotations for the service account created
- use a provided service account

You can check https://github.com/caraml-dev/helm-charts/blob/main/charts/mlp/templates/service-account.yaml for reference

krithika369 replied:
Updated.

terryyylim commented on charts/turing/templates/turing-cluster-role-binding.yaml:
```diff
@@ -0,0 +1,16 @@
+{{- if .Values.clusterConfig.useInClusterConfig -}}
```
Is it necessary for all files in `templates` directory to have `turing` prefix?

krithika369 replied:
Not really, but that's the convention the Merlin templates are following. So I did the same.

terryyylim commented on charts/turing/values.yaml:
```diff
@@ -0,0 +1,300 @@
+# for the detailed explanation on Turing API config options
+# @default -- computed value
+config:
+  AlertConfig:
```
Would be nice to standardise these configs to camelCase instead.

mbruner commented on charts/turing/templates/turing-deployment.yaml:
```diff
@@ -0,0 +1,130 @@
+              key: {{ include "common.postgres-password-secret-key" (list .Values.turingExternalPostgresql) }}
+        {{- else }}
+        - name: DBCONFIG_PASSWORD
+          value: {{ .Values.turingExternalPostgresql.password }}
```
Creating a secret instead of setting a sensitive value in the env variable is better because everyone who can view the manifest will see the password. 

We have a few scenarios possible:
- use in-cluster postgres:
  1. create a secret with password
  2. source env variable from the new secret

- provide the password to external postgres:
  1. create a secret with password
  2. source env variable from the new secret

- provide secret with a password:
  1. source env variable from the secret provided

Currently, we use the third: create secrets outside of the chart and refer them in values

krithika369 replied:
Thanks for the detailed explanation, Maksym. I've updated this.

terryyylim commented on charts/turing/values.yaml:
```diff
@@ -0,0 +1,300 @@
+  password: password
+  # -- Host address for the External postgres
+  address: 127.0.0.1
+turing-postgresql:
```
`turingPostgresql` for consistency?

terryyylim commented on charts/turing/README.md:
```diff
@@ -0,0 +1,73 @@
+# turing
```
Shall we parameterise this file with `README.md.gotmpl` and `helm-docs`?

krithika369 replied:
We don't have this for any of the CaraML components. 🤔 @mbruner any thoughts on this? Shall we add it in?

mbruner replied:
@krithika369 we have it for a few charts `routes`, `common`

krithika369 replied:
Added in the template for Merlin and Turing.

terryyylim reviewed (commented):
Largely LGTM, left some comments, thank you @krithika369 !

terryyylim commented on charts/turing/values.yaml:
```diff
@@ -0,0 +1,300 @@
+
+deployment:
+  image:
+    # -- Docker registry for Turing API image. User is required to override
```
Outdated comment? 🤔

krithika369 replied:
Yes, updated it.

terryyylim commented on charts/turing/templates/_helpers.tpl:
```diff
@@ -0,0 +1,248 @@
+  {{- else if .Values.turingExternalPostgresql.enabled -}}
+    {{- .Values.turingExternalPostgresql.username -}}
+  {{- else -}}
+    {{- .Values.global.postgresqlUsername -}}
```
Just curious, should this be `{{- .Values.global.turing.postgresqlUsername -}}` instead?

krithika369 replied:
No, it is actually meant to be a single set of "global" configs for the DB. It's set up the same way for Merlin.

krithika369 commented:
Hey @mbruner / @terryyylim, I think I've addressed all the comments (and finally got the CI to pass). Please have a look and LMK if you have any other comments. Thanks!

mbruner commented on charts/turing/templates/_helpers.tpl:
```diff
@@ -0,0 +1,282 @@
+Postgres related
+*/}}
+
+{{- define "turing-postgresql.host" -}}
```
maybe using common templates could save some lines of code here?
https://github.com/caraml-dev/helm-charts/blob/main/charts/common/templates/_names.tpl

mbruner commented on charts/turing/templates/turing-config-secret.yaml:
```diff
@@ -0,0 +1,14 @@
+  namespace: {{ .Release.Namespace }}
+  name: {{ template "turing.fullname" . }}-config
+  labels:
+{{ include "turing.labels" . | indent 4 }}
```
You may use `{{- }}` for consistency

mbruner commented on charts/turing/templates/turing-service-account.yaml:
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: {{ template "turing.serviceAccountName" . }}
```
This could be "default" in some cases. Do you think we should still apply this resource?

krithika369 replied:
Removed.

mbruner commented on charts/turing/values.yaml:
```diff
@@ -0,0 +1,314 @@
+  annotations: {}
+    # example.io/example: example
+
+ingress:
```
It is also nice to have annotations for ingress too

krithika369 commented:
Thanks for the review, all. Merging.
